rfcs/003-nix-nickel.md

```diff
@@ -0,0 +1,330 @@
+to calling Nix functions like `override`, `overrideAttrs`, etc (for any override
+that doesn't operate directly on a derivation, which most are). That is, solving
+**LIB** would solve **OVD** as well.
+
+### String contexts
+
+[String
+contexts](https://shealevy.com/blog/2018/08/05/understanding-nixs-string-context/)
+is a very useful feature of Nix. In a Nix program, each string carries a context
+recording which other derivations it depends on at runtime, by tracking
+variables usage inside string interpolation. Nix can then automatically
+determine the dependencies of a derivation.
+
+Abandoning the automatic dependency management offered by string contexts in
+Nickel is an unacceptable degradation of the quality of life of Nix users.
+
+## Proposal
+
+### What could Nix-Nickel look like
+
+Let's imagine what would in practice our ideal way of writing packages in
+Nickel. The main inspiration for this section is Eelco Dolstra's [reflection on
+the Nix language of the future][nix-lang].
+
+#### Package as data
+
+Despite Nix being branded as a _functional_ package manager, a perhaps
+surprising conclusion of Eelco's document is that writing packages as actual
+functions is retrospectively of questionable value. Functions need to be applied
+(and thus arguments be produced) before we can access any data (which hurts
+discoverability), their inputs are hard to override, etc. Overall, functions are
+opaque _computations_ (or _codata_), which makes them hard to inspect and to
+patch. The reflection above pushes to switch to a model where packages are
+rather _data_. Of course, computations still take place -- this is after all the
+whole point of having a configuration language -- but the right representation
+for a package may be better separate and expose pure data and represent
+computations as data dependencies.
+
+Concretely, a derivation would be better represented simply as a recursive
+records:
+
+```nickel
+builders.derivation = {
+
+  # Interface
+
+  name
+    | doc "Name of the derivation, used in the Nix store path."
+    | Str,
+
+  version
+    | doc "Version of the derivation, used in the Nix store path."
+    | Str,
+    | default = "",
+
+  builder
+    | doc "Command to be executed to build the derivation."
+    | Path,
+
+  args
+    | doc "Arguments passed to the builder."
+    | Array Str
+    | default = [],
+
+  outputs
+    | doc "Symbolic names of the outputs of this derivation."
+    | Array Str
+    | default = ["out"],
+
+  env
+    | doc "Structured values passed to the builder."
+    | {_: Str}
+    # inherit doesn't exist in Nickel, but let's pretend, for conciseness
+    = {inherit outputs},
+
+  # Implementation
+
+  drv
+    | doc "The resulting store derivation."
+    =
+    (env & {
+      name = "%{name}-%{version}";
+      inherit builder args;
+    })
+    |> builtin.derivation
+}
+
+# ... other definitions building on this one
+
+# actual package
+pkgs.hello = builders.unix_package & {
```

Comment:
True, and flakes are one example I think, where `output` is still a function but that's fine because the other metadata are one level up and therefore accessible without having to evaluate the output function. And back to our example, evaluating a merge and a simple function (which is morally a parametrized record) shouldn't make a big difference in practice.

One reason I could see to do that would be for consistency, as using both functions and merging may end up being confusing to the user, in particular which one is to be used in which case.